e3sm_diags/driver/zonal_mean_2d_driver.py

```diff
@@ -1,283 +1,347 @@
-import os
-
-import cdms2
-import cdutil
-import MV2
-import numpy
-
-import e3sm_diags
-from e3sm_diags.driver import utils
+import copy
+from typing import List, Tuple
+
+import xarray as xr
+import xcdat as xc  # noqa: F401
+
+from e3sm_diags.driver.utils.dataset_xr import Dataset
+from e3sm_diags.driver.utils.io import _save_data_metrics_and_plots
+from e3sm_diags.driver.utils.regrid import (
+    align_grids_to_lower_res,
+    has_z_axis,
+    regrid_z_axis_to_plevs,
+    subset_and_align_datasets,
+)
+from e3sm_diags.driver.utils.type_annotations import MetricsDict
 from e3sm_diags.logger import custom_logger
-from e3sm_diags.metrics import corr, max_cdms, mean, min_cdms, rmse
-from e3sm_diags.parameter.zonal_mean_2d_parameter import ZonalMean2dParameter
-from e3sm_diags.plot import plot
+from e3sm_diags.metrics.metrics import correlation, rmse, spatial_avg
+from e3sm_diags.parameter.zonal_mean_2d_parameter import (
+    DEFAULT_PLEVS,
+    ZonalMean2dParameter,
+)
+from e3sm_diags.plot.cartopy.zonal_mean_2d_plot import plot as plot_func
 
 logger = custom_logger(__name__)
 
-
-def create_metrics(ref, test, ref_regrid, test_regrid, diff):
-    """Creates the mean, max, min, rmse, corr in a dictionary"""
-    orig_bounds = cdms2.getAutoBounds()
-    cdms2.setAutoBounds(1)
-    lev = ref.getLevel()
-    if lev is not None:
-        lev.setBounds(None)
-
-    lev = test.getLevel()
-    if lev is not None:
-        lev.setBounds(None)
-
-    lev = test_regrid.getLevel()
-    if lev is not None:
-        lev.setBounds(None)
-
-    lev = ref_regrid.getLevel()
-    if lev is not None:
-        lev.setBounds(None)
-
-    lev = diff.getLevel()
-    if lev is not None:
-        lev.setBounds(None)
-    cdms2.setAutoBounds(orig_bounds)
-
-    metrics_dict = {}
-    metrics_dict["ref"] = {
-        "min": min_cdms(ref),
-        "max": max_cdms(ref),
-        "mean": mean(ref, axis="yz"),
-    }
-    metrics_dict["test"] = {
-        "min": min_cdms(test),
-        "max": max_cdms(test),
-        "mean": mean(test, axis="yz"),
-    }
-
-    metrics_dict["diff"] = {
-        "min": min_cdms(diff),
-        "max": max_cdms(diff),
-        "mean": mean(diff, axis="yz"),
-    }
-    metrics_dict["misc"] = {
-        "rmse": rmse(test_regrid, ref_regrid, axis="yz"),
-        "corr": corr(test_regrid, ref_regrid, axis="yz"),
-    }
-
-    return metrics_dict
+DEFAULT_PLEVS = copy.deepcopy(DEFAULT_PLEVS)
 
 
 def run_diag(
-    parameter: ZonalMean2dParameter, default_plevs=ZonalMean2dParameter().plevs
+    parameter: ZonalMean2dParameter, default_plevs=DEFAULT_PLEVS
```

Comment:
I think the expected behavior for default_plevs here should be:
default_plevs = ZonalMean2dParameter().plevs if user specfied it, otherwise to use DEFAULT_PLEVS. I'm not sure if this behavior is being preserved.

Reply:
The current logic is the same as it was before, just the variable name is different (`DEFAULT_PLEVS`):
https://github.com/E3SM-Project/e3sm_diags/blob/3061cecc315396f022c46a4f9f14cf1bf6deb046/e3sm_diags/driver/zonal_mean_2d_driver.py#L68-L70

`default_plevs` is only used if `parameter.plevs` is not set (user specified plevs):

https://github.com/E3SM-Project/e3sm_diags/blob/d0ac18aaaf778bfd5e088b62d81122e09bd58bca/e3sm_diags/driver/zonal_mean_2d_driver.py#L37-L38

Reply:
The reason why we need to pass `default_diags` with a default value is because we wrap `run_diag` in `zonal_mean_2d_stratosphere.py` which has its own `default_diags`.

https://github.com/E3SM-Project/e3sm_diags/blob/refactor/655-zonal-mean-2d/e3sm_diags/driver/zonal_mean_2d_stratosphere_driver.py

I believe this is the only set where we actually have as `default_plevs` arg in `run_diag`.